geo/src/algorithm/bool_ops/mod.rs

```diff
@@ -140,3 +224,27 @@ impl<T: BoolOpsNum> BooleanOps for MultiPolygon<T> {
         self.0.iter().flat_map(|p| p.rings())
     }
 }
+
+/// Allows the unary union operation to be performed on any container which can produce items of type `Polygon<T>`
+impl<T, Container> UnaryUnion for Container
+where
+    T: BoolOpsNum,
+    Container: IntoIterator<Item = Polygon<T>> + Clone,
+    Polygon<T>: RTreeObject,
+{
+    type Scalar = T;
+
+    fn unary_union(&self) -> MultiPolygon<Self::Scalar> {
+        // these three functions drive the union operation
+        let init = || MultiPolygon::<T>::new(vec![]);
+        let fold = |mut accum: MultiPolygon<T>, poly: &Polygon<T>| -> MultiPolygon<T> {
+            accum = accum.union(poly);
+            accum
+        };
+        let reduce = |accum1: MultiPolygon<T>, accum2: MultiPolygon<T>| -> MultiPolygon<T> {
+            accum1.union(&accum2)
+        };
+        let rtree = RTree::bulk_load(self.clone().into_iter().collect());
```

Comment:
I wonder whether it might be more general to require cloneability only of the items instead of the whole container (which might unnecessarily clone auxiliary data structures), e.g.

 ```suggestion
        let rtree = RTree::bulk_load(self.iter().cloned().collect());
```

with the trait bounds adjusted accordingly.

Reply:
Unless I'm misunderstanding you, we would have to restrict the container types to `Vec` and slices if we wanted `iter. Even if the `Container` items impl `Clone`, if we stick with `IntoIter` we have to clone the entire value because it's behind a shared reference:

```
impl<T, Boppable, BoppableCollection> UnaryUnion for BoppableCollection
where
    T: BoolOpsNum,
    Boppable: BooleanOps<Scalar = T> + RTreeObject + Clone,
    BoppableCollection: IntoIterator<Item = Boppable>,
{
    ...
    let rtree = RTree::bulk_load(
        self.into_iter()
            .map(|p| CachedEnvelope::new(p.clone()))
            .collect(),
    );    
}
```

Won't work.

Reply:
I think the trait should avoid imposing the reference and follow `IntoIterator` in being implemented on references instead, e.g. the following diff appears to work:

```diff
diff --git a/geo/src/algorithm/bool_ops/mod.rs b/geo/src/algorithm/bool_ops/mod.rs
index 814b00bb..bf3b898b 100644
--- a/geo/src/algorithm/bool_ops/mod.rs
+++ b/geo/src/algorithm/bool_ops/mod.rs
@@ -168,7 +168,7 @@ pub trait UnaryUnion {
     /// let combined = mp.unary_union();
     /// assert_eq!(&combined, merged);
     /// ```
-    fn unary_union(&self) -> MultiPolygon<Self::Scalar>;
+    fn unary_union(self) -> MultiPolygon<Self::Scalar>;
 }
 
 // This function carries out a full post-order traversal of the tree, building up MultiPolygons from inside to outside.
@@ -228,29 +228,42 @@ impl<T: BoolOpsNum> BooleanOps for MultiPolygon<T> {
     }
 }
 
-impl<T, Boppable, BoppableCollection> UnaryUnion for BoppableCollection
+struct RTreeObjectRef<'a, T>(&'a T);
+
+impl<'a, T> RTreeObject for RTreeObjectRef<'a, T>
+where
+    T: RTreeObject,
+{
+    type Envelope = T::Envelope;
+
+    fn envelope(&self) -> Self::Envelope {
+        self.0.envelope()
+    }
+}
+
+impl<'a, T, Boppable, BoppableCollection> UnaryUnion for &'a BoppableCollection
 where
     T: BoolOpsNum,
-    Boppable: BooleanOps<Scalar = T> + RTreeObject,
-    BoppableCollection: IntoIterator<Item = Boppable> + Clone,
+    Boppable: BooleanOps<Scalar = T> + RTreeObject + 'a,
+    Self: IntoIterator<Item = &'a Boppable>,
 {
     type Scalar = T;
 
-    fn unary_union(&self) -> MultiPolygon<Self::Scalar> {
+    fn unary_union(self) -> MultiPolygon<Self::Scalar> {
         // these three functions drive the union operation
         let init = || MultiPolygon::<T>::new(vec![]);
-        let fold =
-            |mut accum: MultiPolygon<T>, poly: &CachedEnvelope<Boppable>| -> MultiPolygon<T> {
-                accum = accum.union(std::ops::Deref::deref(poly));
-                accum
-            };
+        let fold = |mut accum: MultiPolygon<T>,
+                    poly: &CachedEnvelope<RTreeObjectRef<'a, Boppable>>|
+         -> MultiPolygon<T> {
+            accum = accum.union((&*poly).0);
+            accum
+        };
         let reduce = |accum1: MultiPolygon<T>, accum2: MultiPolygon<T>| -> MultiPolygon<T> {
             accum1.union(&accum2)
         };
         let rtree = RTree::bulk_load(
-            self.clone()
-                .into_iter()
-                .map(|p| CachedEnvelope::new(p))
+            self.into_iter()
+                .map(|p| CachedEnvelope::new(RTreeObjectRef(p)))
                 .collect(),
         );
         bottom_up_fold_reduce(&rtree, init, fold, reduce)
```

and it avoids any cloning at all by building an R tree of short-lived references to the original objects. (I think that `RTreeObjectRef` combinator probably belongs into `rstar` as a blanket impl.)

Reply:
Wow. Yep, that works. We should definitely add this to rstar. If I understand this correctly, it's restricted to situations in which the original collection and its items have the same (`'a`) lifetime as the tree? That's extremely useful.

Reply:
> We should definitely add this to rstar. 

https://github.com/georust/rstar/pull/178

> If I understand this correctly, it's restricted to situations in which the original collection and its items have the same ('a) lifetime as the tree?

The references need to outlive the tree itself. I'd say that this is not really different from creating a `Vec<&'a T>`, we just have to wire through the trait bounds via the wrapper.